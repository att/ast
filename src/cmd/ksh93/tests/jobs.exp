# Interactive tests for jobs, bg, fg and related builtins
set pid [spawn $ksh]
expect_prompt

# ======
# Ctrl-Z should stop current process
log_test_entry
send "cat\r"
sleep 0.5
send [ctrl Z]
expect -re ".*Stopped.*cat\r\n" {
    puts "Ctrl-z stops cat processs"
}
expect_prompt

# ======
# Test if jobs builtin lists stopped process
log_test_entry
send "jobs\r"
expect -re ".*Stopped.*cat\r\n" {
    puts "jobs lists stopped cat process"
}
expect_prompt

# ======
# bg should put last stopped process in background
log_test_entry
send "bg\r"
expect -re ".*cat&\r\n"
expect_prompt

send "jobs\r"
expect -re ".*Stopped.*SIGTTIN.*cat\r\n" {
    puts "bg brings last stopped process to background"
}
expect_prompt

# ======
# fg should put last stopped process in foreground
log_test_entry
send "fg\r"
send "hello world\r\n"
send [ctrl D]
expect -re "\r\nhello world\r\n" {
    puts "fg brings last stopped process to foreground"
}
expect_prompt

# ======
# There should be a warning if shell attempts exit with stopped processes
log_test_entry
send "cat\r"
sleep 0.5
send [ctrl Z]
expect -re ".*Stopped.*cat\r\n"
expect_prompt

# Try to exit shell
send [ctrl D]
expect -re ".*You have stopped jobs\r\n" {
    puts "Warning before exit for stopped jobs works"
}
expect_prompt

# This should actually exit the shell
send [ctrl D]
catch {expect default exp_continue} output
log_debug "EOF output: $output"

catch {wait}
exit 0
