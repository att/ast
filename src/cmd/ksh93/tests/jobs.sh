{ $SHELL -c 'kill %' ;} 2> /dev/null
[[ $? == 1 ]] || log_error "'kill %' has wrong exit status"

wait
unset i
integer i
for (( i=0 ; i < 256 ; i++ ))
do
    sleep 2 &
done

while ! wait
do
    true
done

[[ $(jobs -l) ]] && log_error 'jobs -l should not have any output'

${SHELL} -c 'kill -1 -$$' 2> /dev/null
[[ $(kill -l $?) == HUP ]] || log_error 'kill -1 -pid not working'
${SHELL} -c 'kill -1 -$$' 2> /dev/null
[[ $(kill -l $?) == HUP ]] || log_error 'kill -n1 -pid not working'
${SHELL} -c 'kill -s HUP -$$' 2> /dev/null
[[ $(kill -l $?) == HUP ]] || log_error 'kill -HUP -pid not working'

if kill -L > /dev/null 2>&1
then
    [[ $(kill -l HUP) == "$(kill -L HUP)" ]] || log_error 'kill -l and kill -L are not the same when given a signal name'
    [[ $(kill -l 9) == "$(kill -L 9)" ]] || log_error 'kill -l and kill -L are not the same when given a signal number'
    [[ $(kill -L) == *'9) KILL'* ]] || log_error 'kill -L output does not contain 9) KILL'
fi

unset pid1 pid2
false &
pid1=$!
pid2=$(
    wait $pid1
    (( $? == 127 )) || log_error "job known to subshell"
    print $!
)
wait $pid1
(( $? == 1 )) || log_error "wait not saving exit value"
wait $pid2
(( $? == 127 )) || log_error "subshell job known to parent"
