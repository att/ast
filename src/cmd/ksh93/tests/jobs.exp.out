Ctrl-z stops cat processs
jobs lists stopped cat process
jobs -l lists pid of stopped cat process
jobs -p lists process group of stopped cat process
bg brings last stopped process to background
fg brings last stopped process to foreground
Warning before exit for stopped jobs works
